python/unityagents/environment.py

```diff
@@ -40,141 +44,136 @@ def __init__(self, file_name, worker_id=0,
-        true_filename = os.path.basename(os.path.normpath(file_name))
-        logger.debug('The true file name is {}'.format(true_filename))
-        launch_string = None
-        if platform == "linux" or platform == "linux2":
-            candidates = glob.glob(os.path.join(cwd, file_name) + '.x86_64')
-            if len(candidates) == 0:
-                candidates = glob.glob(os.path.join(cwd, file_name) + '.x86')
-            if len(candidates) == 0:
-                candidates = glob.glob(file_name + '.x86_64')
-            if len(candidates) == 0:
-                candidates = glob.glob(file_name + '.x86')
-            if len(candidates) > 0:
-                launch_string = candidates[0]
-
-        elif platform == 'darwin':
-            candidates = glob.glob(os.path.join(cwd, file_name + '.app', 'Contents', 'MacOS', true_filename))
-            if len(candidates) == 0:
-                candidates = glob.glob(os.path.join(file_name + '.app', 'Contents', 'MacOS', true_filename))
-            if len(candidates) == 0:
-                candidates = glob.glob(os.path.join(cwd, file_name + '.app', 'Contents', 'MacOS', '*'))
-            if len(candidates) == 0:
-                candidates = glob.glob(os.path.join(file_name + '.app', 'Contents', 'MacOS', '*'))
-            if len(candidates) > 0:
-                launch_string = candidates[0]
-        elif platform == 'win32':
-            candidates = glob.glob(os.path.join(cwd, file_name + '.exe'))
-            if len(candidates) == 0:
-                candidates = glob.glob(file_name + '.exe')
-            if len(candidates) > 0:
-                launch_string = candidates[0]
-        if launch_string is None:
-            self.close()
-            raise UnityEnvironmentException("Couldn't launch the {0} environment. "
-                                            "Provided filename does not match any environments."
-                                            .format(true_filename))
-        else:
-            logger.debug("This is the launch string {}".format(launch_string))
-            # Launch Unity environment
-            if docker_training == False:
-                proc1 = subprocess.Popen(
-                    [launch_string,
-                     '--port', str(self.port),
-                     '--seed', str(seed)])
+        self.proc1 = None
+
+        self.communicator = GrpcCommunicator(worker_id, base_port)
+        # self.communicator = SocketCommunicator(worker_id, base_port)
+
+        if file_name != 'editor':
+            cwd = os.getcwd()
+            file_name = (file_name.strip()
+                         .replace('.app', '').replace('.exe', '').replace('.x86_64', '').replace('.x86', ''))
+            true_filename = os.path.basename(os.path.normpath(file_name))
+            logger.debug('The true file name is {}'.format(true_filename))
+            launch_string = None
+            if platform == "linux" or platform == "linux2":
```

Comment:
It's a lot nicer to refactor this code into separate functions. This way, you can easily unit test them, and you don't have huge if-else ladders

Reply:
I intentionally left this code as it is since it is not part of the communicator scope.